### PR TITLE
SANS basics

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Moved and improved from `ess-legacy` repo.

- Combine reduction with and without wavelength bands.
- Avoid hard-coding reduction and grouping dimensions (previously `spectrum` and `layer`).
- Configurable reducer, to support summing all spectra, or summing in groups.
- Use label-based indexing instead of `make_slices` helper.
- Rename `reduce` to `reduce_to_q` since `reduce` shadows a builtin name. Open to better suggestions.

For now this is untried, this will be done as follow-ups when other components get moved.

### CODE REVIEW

**Review by wpotrzebowski — CHANGES_REQUESTED** [review]
1. convert function is from `scipp`not `scippneutron` (old API)
2. `reducer `is not defined here and I presume it is not built-in function
3. I don't quite understand line 30 in reduction.py. Is this, what you meant?

Otherwise, code looks good to me.

File: src/ess/sans/reduction.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    data.coords['wavelength'] = wavelength
+    bands = None
+    for i in range(wavelength_bands.sizes['wavelength'] - 1):
+        low = wavelength_bands['wavelength', i]
+        high = wavelength_bands['wavelength', i + 1]
+        band = sc.histogram(data['wavelength', low:high], q_bins)
+        band = reducer(band)
+        bands = sc.concatenate(bands, band,
+                               'wavelength') if bands is not None else band
+    bands.coords['wavelength'] = wavelength_bands
```
Comment (SimonHeybrock) [inline src/ess/sans/reduction.py:35]: @wpotrzebowski 
> I don't quite understand line 30 in reduction.py. Is this, what you meant?

If you process with wavelength bands, your output will have an additional dimension (`dims=['wavelength', 'Q']`). Here we store the band information as a coord for this extra dim.

**Comment by SimonHeybrock** [comment]
>     1. convert function is from `scipp`not `scippneutron` (old API)
> 
>     2. `reducer `is not defined here and I presume it is not built-in function

See update, hope that explains.

**Review by wpotrzebowski — COMMENTED** [review]
Ok, thanks for explaining. Looks good.  I don't remember this but does `scatter` arg in `convert`  have default True or does it need to be set up explicitly?

**Review by wpotrzebowski — APPROVED** [review]

**Pull request merged**